File: exam/03/lab/binary_bug.cc
// alternative provided by TA 鄧晉杰
struct Trie
{
  struct Node
  {
    long count;
    std::map<char, std::shared_ptr<Node>> children;
  };

  std::shared_ptr<Node> root;

  Trie (): root {std::make_shared<Node>()} {}

  template <typename StringIterator>
  void Insert (StringIterator it, StringIterator end)
  {
    auto node {root};
    while (it != end)
    {
      if (node->children.find(*it) == node->children.end())
      {
        node->children[*it] = std::make_shared<Node>();
      }
      node = node->children[*it];
      ++(node->count);
      ++it;
    }
  }

  template <typename StringIterator>
  long Count (StringIterator it, StringIterator end)
  {
    auto node {root};
    while (it != end)
    {
      if ((node != nullptr) && (node->children.find(*it) != node->children.end()))
      {
        node = node->children[*it];
      }
      else
      {
        return 0;
      }
      ++it;
    }
    return node->count;
  }
};

std::vector<long> CountBug (long n, std::vector<std::string> C, long m, std::vector<std::string> B)
{
  std::vector<long> responses;
  Trie trie;
  for (auto &code : C)
  {
    for (auto it {code.begin()}; it != code.end(); ++it)
    {
      trie.Insert(it, code.end());
    }
  }
  for (auto &bug : B)
  {
    responses.push_back(trie.Count(bug.begin(), bug.end()));
  }
  return responses;
}

File: exam/03/lab/binary_bug.py
# alternative provided by TA 鄧晉杰
class Trie:
    class Node:
        def __init__ (self):
            self.count = 0
            self.children = {}

    def __init__ (self):
        self.root = self.Node()

    def Insert (self, suffix):
        node = self.root
        for character in suffix:
            if character not in node.children:
                node.children[character] = self.Node()
            node = node.children[character]
            node.count += 1

    def Count (self, string):
        node = self.root
        for character in string:
            if character not in node.children:
                return 0
            node = node.children[character]
        return node.count

def CountBug (n, C, m, B):
    responses = []
    trie = Trie()
    for code in C:
        for i in range(len(code)):
            trie.Insert(code[i:])
    for bug in B:
        responses.append(trie.Count(bug))
    return responses
